vendor/github.com/ethereum/go-ethereum/eth/downloader/downloader.go

```diff
@@ -476,12 +479,17 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td *big.I
 	} else if d.mode == FullSync {
 		fetchers = append(fetchers, d.processFullSyncContent)
 	}
-	return d.spawnSync(fetchers)
+	return d.spawnSync(errCancelHeaderFetch, fetchers)
 }
 
 // spawnSync runs d.process and all given fetcher functions to completion in
 // separate goroutines, returning the first error that appears.
-func (d *Downloader) spawnSync(fetchers []func() error) error {
+func (d *Downloader) spawnSync(errCancel error, fetchers []func() error) error {
+	select {
+	case <-d.cancelCh:
+		return errCancel
+	default:
+	}
```

Comment:
from my pov this is the only place that needs to be changed in this patch

Reply:
Let's try removing this one, since we've introduced a check further up in the same method. Maybe it's not required anymore. That way this commit will match exactly the one in the upstream PR.

Reply:
i dont follow, why to remove it? this is exactly where race condition happened

Reply:
or that check wasn't in place when error happened?

Reply:
sorry, I meant to write "the other one", not "this one" :-/